File: findTheTownJudge.java
//TC: O(n)
//SC: O(n) 

class Solution {
    public int findJudge(int N, int[][] trust) {

        int[] inDegree = new int[N+1];
        int[] outDegree = new int[N+1];
        
        for(int i = 0;i < trust.length; i++){
            inDegree[trust[i][1]]+=1;
            outDegree[trust[i][0]]+=1;
        }
        
        //check the conditions for becoming town judge:
        //1. trusted by N-1 people
        //2. not having back edge
        
        for(int i =1;i<inDegree.length;i++){
            if(inDegree[i] == (N-1) && outDegree[i] == 0){
                return i;
            }
        }
        
        return -1;
    }
}

File: Find the Town Judge.cpp
//Time Complexity- O(E) //number of edges
//Space Complexity- O(n) //number of people

class Solution {
public:
    int findJudge(int N, vector<vector<int>>& trust) {
        
        vector<int> trusts(N+1,0);
        vector<int> trustedby(N+1,0);
        
        for(int i=0;i<trust.size();i++){
            trusts[trust[i][0]]++;
            trustedby[trust[i][1]]++;
        }
        
        int answer=-1;
        for(int i=1;i<=N;i++){
            if(trusts[i]==0 && trustedby[i]==N-1){
                answer=i;
                break;
            }
        }
        return answer;
    }
};